src/app/loadout-builder/mod-utils.ts

```diff
@@ -31,3 +41,299 @@ export function someModHasEnergyRequirement(mods: PluggableInventoryItemDefiniti
+      );
+      itemForMod && bucketSpecificAssignments.get(itemForMod.id)?.push(mod);
+    }
+  }
+
+  const itemEnergies = _.mapValues(
+    _.keyBy(items, (item) => item.id),
+    (item) => ({
+      used: _.sumBy(
+        bucketSpecificAssignments.get(item.id),
+        (mod) => mod.plug.energyCost?.energyCost || 0
+      ),
+      originalCapacity: item.energy?.energyCapacity || 0,
+      derivedCapacity: upgradeSpendTierToMaxEnergy(defs, upgradeSpendTier, item),
+      originalType: item.energy?.energyType || DestinyEnergyType.Any,
+      derivedType: getItemEnergyType(
+        defs,
+        item,
+        upgradeSpendTier,
+        lockItemEnergyType,
+        bucketSpecificAssignments[item.id]
+      ),
+    })
+  );
+
+  const generalModPermutations = generatePermutationsOfFive(generalMods, stringifyMods);
+  const combatModPermutations = generatePermutationsOfFive(combatMods, stringifyMods);
+  const raidModPermutations = generatePermutationsOfFive(raidMods, stringifyMods);
+
+  combatModLoop: for (const combatP of combatModPermutations) {
+    combatItemLoop: for (let i = 0; i < items.length; i++) {
+      const combatMod = combatP[i];
+
+      // If a mod is null there is nothing being socketed into the item so move on
+      if (!combatMod) {
+        continue combatItemLoop;
+      }
+
+      const item = items[i];
+      const itemEnergy = itemEnergies[item.id];
+      const modTag = getModTypeTagByPlugCategoryHash(combatMod.plug.plugCategoryHash);
+      const combatEnergyCost = combatMod.plug.energyCost?.energyCost || 0;
+      const combatEnergyType = combatMod.plug.energyCost?.energyType || DestinyEnergyType.Any;
+
+      const combatEnergyIsValid =
+        itemEnergy &&
+        itemEnergy.used + combatEnergyCost <= itemEnergy.derivedCapacity &&
+        energyTypesAreCompatible(itemEnergy.derivedType, combatEnergyType);
+
+      // The other mods wont fit in the item set so move on to the next set of mods
+      if (
+        !(
+          combatEnergyIsValid &&
+          modTag &&
+          itemSocketMetadata[item.id]?.some((metadata) =>
+            metadata.compatibleModTags.includes(modTag)
+          )
+        )
+      ) {
+        continue combatModLoop;
+      }
+    }
+
+    generalModLoop: for (const generalP of generalModPermutations) {
+      generalItemLoop: for (let i = 0; i < items.length; i++) {
+        const generalMod = generalP[i];
+
+        // If a mod is null there is nothing being socketed into the item so move on
+        if (!generalMod) {
+          continue generalItemLoop;
+        }
+
+        const item = items[i];
+        const itemEnergy = itemEnergies[item.id];
+        const generalEnergyCost = generalMod.plug.energyCost?.energyCost || 0;
+        const generalEnergyType = generalMod.plug.energyCost?.energyType || DestinyEnergyType.Any;
+        const combatEnergyCost = combatP?.[i]?.plug.energyCost?.energyCost || 0;
```

Comment:
Is combatP in scope here? I can't quite tell if `generalModLoop` is nested inside `combatModLoop`

Reply:
Yes it is, this is a three tier loop as we need to check the validity of all three mods energy type at the final phase. I was hoping to un-nest them but I haven't found a way to do so yet.

Reply:
I will put this in a comment but the hierarchy looks like this
```
combatModLoop
--combatItemLoop
--generalModLoop 
----generalItemLoop
----raidModLoop
------raidItemLoop
```